Analysis plan
=============
This analysis pipeline aims to identify the genomic loci that affect the expression of genes in different tumor types.

## 1. Prepare data for QC:
### 1.1 Transfer BIRDSEED format to VCF format
In total 22936 Affymetrix Genome-Wide Human SNP Array 6.0 platform SNParray data from all 11161 patients from TCGA database were downloaded in BIRDSEED format, which were then transferred to VCF format using [BirdSuite] (https://www.broadinstitute.org/birdsuite/birdsuite-downloads) developed by Broad Institute. For the source code in file BirdseedToVCF.py, you need to comment out line 270 and 279 to avoid errors. To run the birdseed, we need to use a reference map file 
[GenomeWideSNP_6.na30.annot.hg19.csv.pickle](https://software.broadinstitute.org/cancer/cga/sites/default/files/data/tools/contest/GenomeWideSNP\_6.na30.annot.hg19.csv.pickle.gz) to map the snp id.

### 1.2 Change header of VCF files.
The VCF files generated by BIRDSEED miss some information in the header, this would cause some problem in the downstream analysis. So you need to change the VCF header to the one provided in the file utils/header.txt using **bcftools reheader**. The header file includes the GT, DP, GQ for FORMAT id and PID for INFO id. Then merge all the VCF files into one using **bcftools merge**. Since there are over 20k VCF files, merging all the files at once would need too much memory, so it's recommended to split the files into chunks (eg: each chunk with 1000 VCF files), first merge files in each chunk then merge the VCF files from each chunk together.

### 1.3 Transfer VCF to PLINK format
VCF file is transferred to .bed, .bim, .fam format (this is plink1 format).

## 3. QC for the variants. 
Remove samples with call rate < 95%, and minor allele frequency < 0.5, also remove SNPs out of Hardy Weinberg equilibrium with p<10-6.
Apply PCA for all the samples using PLINK and select Caucasian samples.
